### ISSUE TITLE
LightningDataModule and Protocol Compatibility

### ISSUE THREAD

**Issue by mogwai** [pyannote/pyannote-audio#483]
<a href="https://github.com/mogwai"><img src="https://avatars1.githubusercontent.com/u/5440566?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [mogwai](https://github.com/mogwai)**
_Saturday Oct 31, 2020 at 17:15 GMT_
_Originally opened as https://github.com/hbredin/pyannote-audio-v2/issues/56_

----

Create a way to convert LightningDataModules into Protocols, demonstrate this with a custom DataModule.

**Comment by mogwai** [comment]
<a href="https://github.com/hbredin"><img src="https://avatars2.githubusercontent.com/u/1080837?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [hbredin](https://github.com/hbredin)**
_Sunday Nov 01, 2020 at 06:42 GMT_

----

Might be worth looking at https://github.com/tmbdev/webdataset

**Comment by mogwai** [comment]
<a href="https://github.com/hbredin"><img src="https://avatars2.githubusercontent.com/u/1080837?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [hbredin](https://github.com/hbredin)**
_Sunday Nov 01, 2020 at 21:17 GMT_

----

I guess this is related to https://github.com/pyannote/pyannote-database/issues/61 as well.

The `LightningDataModules` could simply prepare its very own `database.yml` file and *register* (whatever that means) it with `pyannote.database`.  Instead of switching `database.yml` with proposed `set_database_yml`, we could register multiple `database.yml` files. I think `add_custom_protocols` already does most of this (it adds keys to `DATABASES` and `TASKS` dictionaries).  

What do you think?

**Comment by mogwai** [comment]
<a href="https://github.com/mogwai"><img src="https://avatars1.githubusercontent.com/u/5440566?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mogwai](https://github.com/mogwai)**
_Monday Nov 02, 2020 at 16:00 GMT_

----

WebDataset is interesting. I'll experiment with generating a protocol programatically. At the moment I think we need the datasets present in: https://github.com/pyannote/pyannote-audio/tree/develop/tutorials/data_preparation and probably voxceleb?

**Comment by mogwai** [comment]
<a href="https://github.com/hbredin"><img src="https://avatars2.githubusercontent.com/u/1080837?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [hbredin](https://github.com/hbredin)**
_Monday Nov 02, 2020 at 20:02 GMT_

----

Yes, AMI, MUSAN and VoxCeleb. Would be nice to have VoxConverse also.

Note that, even though VoxCeleb and VoxConverse are free to download, one first needs to get a login and password from the data provider.

**Comment by mogwai** [comment]
<a href="https://github.com/mogwai"><img src="https://avatars1.githubusercontent.com/u/5440566?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mogwai](https://github.com/mogwai)**
_Monday Nov 02, 2020 at 21:39 GMT_

----

I found a download script for them so we'll just need to user to configure some enviroment variables with the user name and password given to them by the websites.

**Comment by mogwai** [comment]
<a href="https://github.com/mogwai"><img src="https://avatars1.githubusercontent.com/u/5440566?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mogwai](https://github.com/mogwai)**
_Tuesday Nov 03, 2020 at 14:22 GMT_

----

I've look into WebDataset and although I think it looks really really promising, it might be a todo for the future. I'd say we follow this issue and wait for it to be added to pytorch before we adopt it? https://github.com/pytorch/pytorch/issues/38419.

For now I think if we focus on creating the conversions between protocols and datamodules, this won't be too tricky to implement on top of that.

**Comment by mogwai** [comment]
<a href="https://github.com/mogwai"><img src="https://avatars1.githubusercontent.com/u/5440566?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mogwai](https://github.com/mogwai)**
_Wednesday Nov 04, 2020 at 13:42 GMT_

----

Just a progress update on this. Yesterday I spent quite a long time trying to figure out a way to add a memory based approach on top of the file based approach without making changes to pyannote.database. I eventually realised that this wasn't going to be easy as the `FileFinder` unavoidable reads the Databases globs from the database.yml file. I'm building a class that will encapsulate the Protocol class, presenting a similar API to make it compatible with the current Tasks.  

The main goals I've laid out so far are:

- Add that dataset to the protocol.yml
- Downloads the data, if not already present
- Extract the data
- Validates (MD5, Loads, normalize)
- Clean (Remove Silence, programmable after functions)

All of these can be configured by force to happen.

The power of this is that we can track changes to the dataset in code. So if we discover the labelling is incorrect somewhere, we can add a function to correct that later, or that a file is corrupted somehow we can remove it from the list of items to download. We can extend the functionality of this as needed, including extending it to use the WebDataset. 

After trying an approach to use the LightningDataModule class but realised that we do this in the Task so there isn't really any benefit to using it for this specific thing. All this class needs to do is provide a list of audio files that are annotated, the tasks will chunk them up as required. 

I do feel that a python first approach in memory that is not locked to a file might be more flexible as we can address issues within a dataset with new releases of our code. It also means we can update the users environment without them having to remove or change files in their system.
